File: src/index/vector-storage.ts
import { VectorIndex } from "./dense.ts";
import { QuantizedVectorIndex } from "./quantize.ts";
import type { SemanticIndex } from "./semantic-index.ts";

export type SemanticStorage = "int8" | "float32";

/**
 * Production default: int8 quantized vectors (~4x less RAM than float32).
 *
 * Set `MIRU_FLOAT_VECTORS=1` to build/load float32 instead. That path exists for
 * debugging, precision baselines, and quantization A/B tests — not normal use.
 */
export function resolveSemanticStorage(): SemanticStorage {
  return process.env.MIRU_FLOAT_VECTORS === "1" ? "float32" : "int8";
}

export function semanticStorageFromMetadata(metadata: Record<string, unknown>): SemanticStorage {
  return metadata.vector_storage === "float32" ? "float32" : "int8";
}

export function buildSemanticIndex(vectors: Float32Array[]): SemanticIndex {
  if (resolveSemanticStorage() === "int8") {
    return new QuantizedVectorIndex(vectors);
  }
  return new VectorIndex(vectors);
}

export function semanticStorageOf(index: SemanticIndex): SemanticStorage {
  return index instanceof QuantizedVectorIndex ? "int8" : "float32";
}

File: internal/index/vector_storage.go
package index

import "os"

type SemanticIndex interface {
	Size() int
	Dimensions() int
	MemoryBytes() int
	VectorAt(docIndex int) ([]float32, error)
	Query(queryVector []float32, k int, selector ...[]int) (QueryResult, error)
}

type SemanticStorage string

const (
	SemanticInt8    SemanticStorage = "int8"
	SemanticFloat32 SemanticStorage = "float32"
)

func ResolveSemanticStorage() SemanticStorage {
	if os.Getenv("MIRU_FLOAT_VECTORS") == "1" || os.Getenv("SEMBLE_FLOAT_VECTORS") == "1" {
		return SemanticFloat32
	}
	return SemanticInt8
}

func SemanticStorageFromMetadata(metadata map[string]any) SemanticStorage {
	if metadata["vector_storage"] == "float32" {
		return SemanticFloat32
	}
	return SemanticInt8
}

func BuildSemanticIndex(vectors [][]float32) SemanticIndex {
	if ResolveSemanticStorage() == SemanticInt8 {
		return NewQuantizedVectorIndex(vectors)
	}
	return NewVectorIndex(vectors)
}

func SemanticStorageOf(index SemanticIndex) SemanticStorage {
	if _, ok := index.(*QuantizedVectorIndex); ok {
		return SemanticInt8
	}
	return SemanticFloat32
}
